examples/VB_ASR/helpers.py

```diff
@@ -0,0 +1,217 @@
+                        ha="center",
+                        va="bottom",
+                        fontsize=sz,
+                    )
+                else:
+                    plt.text(
+                        t,
+                        0.9 * i * max_pdf / (len(target) - 1),
+                        f"target{i}",
+                        ha="center",
+                        va="bottom",
+                        fontsize=sz,
+                    )
+            t_min, t_max = min(target), max(target)
+        else:
+            target_ = np.format_float_scientific(target, unique=False, precision=2)
+            plt.bar(
+                x=target,
+                height=1.0 * max_pdf,
+                width=max_std * 0.03,
+                color=PltOptions["perfect"]["color"],
+            )
+            if _show_values:
+                plt.text(
+                    target,
+                    0,
+                    f"target={target_}",
+                    ha="center",
+                    va="bottom",
+                    fontsize=sz,
+                )
+            else:
+                plt.text(target, 0, f"target", ha="center", va="bottom", fontsize=sz)
+            t_min, t_max = target, target
+    plt.xticks(np.linspace(min(x_min, t_min), max(x_max, t_max), 3), fontsize=sz)
+    plt.yticks(np.linspace(0.0, max_pdf, 3), fontsize=sz)
+    plt.title(tit, fontsize=sz)
+    plt.xlabel(xl, fontsize=sz)
+    plt.ylabel(yl, fontsize=sz)
+    plt.legend(fontsize=int(sz * 0.7), loc="best")
+    plt.tight_layout()
+    plt.savefig(_path + _name + _format, bbox_inches="tight", dpi=dpi)
+    plt.show(block=False)
+    plt.rcParams["text.usetex"] = False
+
+
+def gaussian_error_propagator(jac, stds=None, cov=None, rtol=1e-12, atol=1e-6):
+    """
+    If we have some random parameters (like 'p') as inputs to a function like 'f':
+    jac:
+        the jacobian (derivative) of 'f' at p_mean (means of 'p'); i.e.:
+            jac = d [f(p_mean)] / d [p_mean] .
+    stds (a vector):
+        standard deviations of 'p'.
+    cov (if not None):
+        the covariance of 'p'. --> its diagonal is stds**2 .
+
+    --> At least one of 'stds' and 'cov' must be given.
+
+    Returns:
+        the full covariance matrix of f(p) .
+
+    NOTEs:
+    - The mean values of p; i.e. p_mean, do not matter!
+    - The tolerances are meant for:
+        - checking if stds and cov are in agreement (if both are given).
+        - checking 'cov' to be symmetric, which is a requirement for the validity of the formula used.
+    """
+    if (stds is None) and (cov is None):
+        raise ValueError(f"At least one of stds or cov must be given.")
+    else:  # At least one of stds and/or cov is given.
+        if (stds is not None) and (
+            cov is not None
+        ):  # if both 'stds' 'cov' are given, we check them to agree.
+            e = np.linalg.norm(np.diag(cov) ** (0.5) - stds) / np.linalg.norm(stds)
```

Comment:
Don't you have to check the offdiagonal elements as well, so rather convert the stds to a diagonal matrix and then check the full matrix.

Reply:
I have used this method such that the stds actually come from the diagonal of the cov. matrix. However, your point is valid and I'll change it.